File: solutions/python/recursion/coin_change.py
CANNOT_PRODUCE_SPECIFIED_AMOUNT = -1

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0

        fewest_coins_to_make_amount = float('inf')

        for coin_amount in coins:
            remaining_amount = amount - coin_amount

            if remaining_amount < 0:
                continue

            remaining_coins_required = self.coinChange(coins, remaining_amount)

            if remaining_coins_required == CANNOT_PRODUCE_SPECIFIED_AMOUNT:
                continue

            fewest_coins_to_make_amount = min(fewest_coins_to_make_amount, 1 + remaining_coins_required)

        if fewest_coins_to_make_amount == float('inf'):
            return CANNOT_PRODUCE_SPECIFIED_AMOUNT

        return fewest_coins_to_make_amount

File: solutions/javascript/recursion/coinChange.js
const CANNOT_PRODUCE_SPECIFIED_AMOUNT = -1;

const coinChange = (coins, amount) => {
    if (amount === 0) return 0;

    let fewestCoinsToMakeAmount = Infinity;

    for (const coinAmount of coins) {
        const remainingAmount = amount - coinAmount;

        if (remainingAmount < 0) continue;

        const remainingCoinsRequired = coinChange(coins, remainingAmount);

        if (remainingCoinsRequired === CANNOT_PRODUCE_SPECIFIED_AMOUNT) continue;

        fewestCoinsToMakeAmount = Math.min(fewestCoinsToMakeAmount, 1 + remainingCoinsRequired);
    }

    if (fewestCoinsToMakeAmount === Infinity) {
        return CANNOT_PRODUCE_SPECIFIED_AMOUNT;
    }

    return fewestCoinsToMakeAmount;
};
